src/consensus/read_validator.rs

```diff
@@ -0,0 +1,145 @@
+use std::collections::BTreeMap;
+
+use crate::core::types::SnapchainValidatorContext;
+use crate::proto::{self, DecidedValue, Height};
+use crate::storage::store::engine::{BlockEngine, ShardEngine};
+use bytes::Bytes;
+use informalsystems_malachitebft_sync::RawDecidedValue;
+use prost::Message;
+use tracing::info;
+
+use super::malachite::read_sync::{self, ReadSyncRef};
+
+pub enum Engine {
+    ShardEngine(ShardEngine),
+    BlockEngine(BlockEngine),
+}
+pub struct ReadValidator {
+    pub engine: Engine,
+    pub shard_id: u32,
+    pub last_height: Height,
+    pub max_num_buffered_blocks: u32,
+    pub buffered_blocks: BTreeMap<Height, proto::DecidedValue>,
+}
+
+impl ReadValidator {
+    pub fn initialize_height(&mut self, sync: &ReadSyncRef) {
+        let height = match &self.engine {
+            Engine::BlockEngine(engine) => engine.get_confirmed_height(),
+            Engine::ShardEngine(engine) => engine.get_confirmed_height(),
+        };
+        self.last_height = height;
+        sync.cast(read_sync::Msg::Decided(height)).unwrap();
+    }
+
+    pub fn get_min_height(&self) -> Height {
+        // Always return the genesis block, until we implement pruning
+        Height::new(self.shard_id, 1)
+    }
+
+    fn commit_decided_value(&mut self, value: &DecidedValue, height: Height, sync: &ReadSyncRef) {
+        match &mut self.engine {
+            Engine::ShardEngine(shard_engine) => match &value.value {
+                Some(proto::decided_value::Value::Shard(shard_chunk)) => {
+                    shard_engine.commit_shard_chunk(&shard_chunk);
+                    info!(
+                        %height,
+                        hash = hex::encode(&shard_chunk.hash),
+                        "Processed decided shard chunk"
+                    );
+                }
+                _ => {
+                    panic!("Invalid decided value")
+                }
+            },
+            Engine::BlockEngine(block_engine) => match &value.value {
+                Some(proto::decided_value::Value::Block(block)) => {
+                    block_engine.commit_block(&block);
+                    info!(
+                        %height,
+                        hash = hex::encode(&block.hash),
+                        "Processed decided block"
+                    );
+                }
+                _ => {
+                    panic!("Invalid decided value")
+                }
+            },
+        };
+        self.last_height = height;
+        sync.cast(read_sync::Msg::Decided(height)).unwrap();
+    }
+
+    fn process_buffered_blocks(&mut self, sync: &read_sync::ReadSyncRef) {
+        while let Some((height, value)) = self.buffered_blocks.pop_first() {
```

Comment:
This is guaranteed to be sorted by height?

Reply:
Yeah, the BTreeMap is sorted by key. I added a comment to make it clear that this is important and prompt us to think before changing.